starlite/contrib/repository/testing/generic_mock_repository.py

```diff
@@ -0,0 +1,346 @@
+"""A repository implementation for tests.
+
+Uses a `dict` for storage.
+"""
+from __future__ import annotations
+
+from datetime import datetime
+from typing import TYPE_CHECKING, Generic, TypeVar
+from uuid import uuid4
+
+from starlite.contrib.repository.abc import AbstractRepository
+from starlite.contrib.repository.exceptions import ConflictError, RepositoryError
+
+if TYPE_CHECKING:
+    from collections.abc import Callable, Hashable, Iterable, MutableMapping
+    from typing import Any
+
+    from starlite.contrib.repository import FilterTypes
+    from starlite.contrib.sqlalchemy import base
+
+ModelT = TypeVar("ModelT", bound="base.Base | base.AuditBase")
+MockRepoT = TypeVar("MockRepoT", bound="GenericMockRepository")
+
+
+class GenericMockRepository(AbstractRepository[ModelT], Generic[ModelT]):
+    """A repository implementation for tests.
+
+    Uses a :class:`dict` for storage.
+    """
+
+    collection: MutableMapping[Hashable, ModelT]
+    model_type: type[ModelT]
+
+    def __init__(self, id_factory: Callable[[], Any] = uuid4, **_: Any) -> None:
+        super().__init__()
+        self._id_factory = id_factory
+
+    @classmethod
+    def __class_getitem__(cls: type[MockRepoT], item: type[ModelT]) -> type[MockRepoT]:
+        """Add collection to ``_collections`` for the type.
+
+        Args:
+            item: The type that the class has been parametrized with.
+        """
+        return type(  # pyright:ignore
+            f"{cls.__name__}[{item.__name__}]", (cls,), {"collection": {}, "model_type": item}
+        )
+
+    def _find_or_raise_not_found(self, item_id: Any) -> ModelT:
+        return self.check_not_found(self.collection.get(item_id))
+
+    async def add(self, data: ModelT, allow_id: bool = False) -> ModelT:
+        """Add ``data`` to the collection.
+
+        Args:
+            data: Instance to be added to the collection.
+            allow_id: disable the identified object check.
+
+        Returns:
+            The added instance.
+        """
+        if allow_id is False and self.get_id_attribute_value(data) is not None:
+            raise ConflictError("`add()` received identified item.")
+        now = datetime.now()  # noqa: DTZ005
+        if hasattr(data, "updated") and hasattr(data, "created"):
+            # maybe the @declarative_mixin decorator doesn't play nice with pyright?
+            data.updated = data.created = now  # pyright: ignore
+        if allow_id is False:
+            id_ = self._id_factory()
+            self.set_id_attribute_value(id_, data)
+        self.collection[data.id] = data
+        return data
+
+    async def add_many(self, data: list[ModelT], allow_id: bool = False) -> list[ModelT]:
```

Comment:
I know this is not your design decision, but `allow_id` shouldn't be in the signature here I don't think as this should be called by production code where the param cannot be set. Maybe we should make it an arg to `__init__()`